### DISCUSSION TITLE
Switching between multiple graphs recommendation?

### DISCUSSION THREAD

**Discussion by joey101** [patrykandpatrick/vico#866] [Questions]
### Question

Hi 

I was wondering if there are any recommendations for having multiple dynamic graphs? My goal is to have a drop down menu and depending on what is chosen it will display a different graph.

Should i have 3 different CartesianChartModelProducer()'s. or should i have one and runTransaction each time with the data specific to what was chosen in the drop down menu to produce the graph that I want to display in that moment?

### Vico version(s)

2.0.0-alpha.28

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hello, @joey101. It depends on how much the three charts differ. Could you share some details on that?

**Comment by Praveen-Android** [comment]
**Thought would share what I am doing for something similar and see if its the way to go since am not certain if its the optimal way to updating graphs or best place to host modelProducer and others that the graph requires (or may be in a view model)**

Have three tabs (month/week/day) and switching between tabs should show/update graphs with aggregated data points over the month/week/day. I am using `tabState.id` as key to run `LaunchedEffect` which is going to update `modelProducer` with new line series.

```
@Composable
fun Chart(
    modifier: Modifier = Modifier,
    tabState: TabState
) {
    val modelProducer = remember { CartesianChartModelProducer() }
    val xToDateMapKey = remember { ExtraStore.Key<Map<Float, LocalDate>>() }


    LaunchedEffect(tabState.id) {
        withContext(Dispatchers.Default) {
            val xToDates = tabState.keys.associateBy { it.toEpochDay().toFloat() }
            modelProducer.runTransaction {
                lineSeries { series(xToDates.keys, tabState.values) }
                extras { it[xToDateMapKey] = xToDates }
            }
        }
    }
    .....
    ShowChart(modifier, modelProducer, xToDateMapKey)
}
```

**Reply by Gowsky** [reply]
Hello, @Praveen-Android. This approach is correct.

> am not certain if its the optimal way to updating graphs or best place to host modelProducer and others that the graph requires (or may be in a view model)

When possible, it's preferable to keep the `CartesianChartModelProducer` in the `ViewModel` and feed it with data there. (It's a poor idea to just lift the `CartesianChartModelProducer` and still run `Transaction`s from the composable — these two things should generally be at the same level.)

Remember that any `ExtraStore.Key`s must be persisted no shorter than their `CartesianChartModelProducer`. Our recommendation for scalability is to create singleton `ExtraStore.Key` objects and store them in static contexts: at the root of a file, in an `object`, etc. This way, you have no risk of incorrectly recreating `ExtraStore.Key`s, and you don't have to pass them around as arguments.
